File: morning_exercises/day26.ts
function createGraph(num: number, prerequisites: number[][]): number[][] {
    let graph:number[][] = Array.from({length: num}, () => []);
    for (const [a,b] of prerequisites) graph[b].push(a);
    return graph;
}

function dfsCheck(graph: number[][], node: number, state: number[]): boolean {
    if (state[node]==2) return true;
    if (state[node]==1) return false;

    state[node]=1;
    for (const neighbour of graph[node]) {
        if (!dfsCheck(graph, neighbour, state)) return false;
    } state[node]=2;
    return true;
}

function isDoable(graph: number[][]): boolean {
    let state: number[] = Array.from({length : graph.length}, () => 0);
    for (let i=0; i<graph.length; i++) {
        if (!dfsCheck(graph, i, state)) return false;
    } return true;
}

{
    let graph = createGraph(4, [[1,0], [2,1], [3,0]]);

    console.log(isDoable(graph));
}

File: morning_exercises/day26.cs
using System.Runtime.CompilerServices;

class Doable
{
    public static List<List<int>> CreateGraph(int number, List<(int, int)> prerequisites)
    {
        List<List<int>> graph = [];
        for (int i = 0; i < number; i++) graph.Add([]);

        foreach (var(a,b) in prerequisites) graph[b].Add(a);
        return graph;
    }

    public static bool CheckDfs(List<List<int>> graph, int node, List<int> state)
    {
        if (state[node]==2) return true;
        if (state[node]==1) return false;

        state[node]=1;
        foreach (int neighbour in graph[node]) if (!CheckDfs(graph, neighbour, state)) return false;
        state[node]=2;
        return true;
    }

    public static bool IsDoable(List<List<int>> graph)
    {
        List<int> state = [];
        for (int i = 0; i < graph.Count; i++) state.Add(0);
        for (int i=0;i<graph.Count; i++) if (!CheckDfs(graph, i, state)) return false;
        return true;
    }
}